### PR TITLE
FoldLeft introduction

### PR DESCRIPTION
**Pull request by ggarvanese**

### CODE REVIEW

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 1]
```diff
@@ -0,0 +1,277 @@
+= Parcourir des collections avec foldLeft et foldRight
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:1]: Remarque générale, si tu ne laisses pas une ligne blanche entre deux lignes, cela ne passe pas à la ligne dans le rendu final comme tu peux voir ici : https://github.com/lunatech-labs/lunatech-blog/blob/foldLeft-introduction/posts/2022-06-27-foldleft-introduction.adoc
Je te conseille d'aérer un peu les paragraphes pour que ça fait moins des gros blocs ;)

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+def foldLeft[B](z: B)(op: (B, A) => B): B
+----
+Nous voyons ici plusieurs choses intéressantes :
+1. La fonction est typée B, qui est le type de sortie
+2. Elle prend en entrée un premier paramètre z (zéro) de ce fameux type B. Ce paramètre est important car c'est le point de départ de la fonction, il sert à l'initialiser et à collecter les résultats. Ce sera notre accumulateur
+3. Et en second paramètre "op" (opération) une fonction qui prend deux paramètres (B, le type de sortie et A , le type des éléments de la liste) et qui permet d'obtenir ce fameux type de sortie B.
+En gros, cette fonction va parcourir tous les éléments de la liste [A] et va collecter le résultat dans l'accumulateur [B]. A la fin, on se retrouve avec [B].
+Commençons par un exemple simple. Nous allons additionner chaque chiffre d'une liste avec son prédécesseur pour obtenir le résultat final.
+
+image::foldleft-decomposition.png[FoldLeft decomposition, width = 700]
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:22]: Sympa les illustrations :)

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+:tags: [foldleft, iteration, functionnal, scala]
+Il y a quelques temps, j'ai été chargé de développer un petit algorithme pour mon projet. Il s'agissait de parcourir toutes les images d'une séquence vidéo afin de déterminer quel label apposé manuellement occupait le plus d'images au sein de cette séquence. Cet exercice m'a permis de (re)découvrir une fonction que je n'utilise que trop rarement au quotidien, le foldLeft (et son corollaire, le foldRight). 
+Ces méthodes, qui permettent de parcourir des listes effectuant des opération entre la valeur courante et la valeur précédente tout en permettant de retourner un type complètement différent, méritent qu'on s'y attarde un peu.
+Intéressons-nous tout d'abord à la signature de cette méthode
+[source, scala]
+----
+def foldLeft[B](z: B)(op: (B, A) => B): B
+----
+Nous voyons ici plusieurs choses intéressantes :
+1. La fonction est typée B, qui est le type de sortie
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:16]: Il faut que tu sautes des lignes entre chaque point sinon tout est sur la même ligne

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 4]
```diff
@@ -0,0 +1,277 @@
+= Parcourir des collections avec foldLeft et foldRight
+Guillaume Garvanèse
+v1.0, 2021-06-27
+:title: Parcourir des collections avec foldLeft et foldRigh
+:imagesdir: ../media/2022-06-27-foldleft-introduction
+:lang: fr
+:tags: [foldleft, iteration, functionnal, scala]
+Il y a quelques temps, j'ai été chargé de développer un petit algorithme pour mon projet. Il s'agissait de parcourir toutes les images d'une séquence vidéo afin de déterminer quel label apposé manuellement occupait le plus d'images au sein de cette séquence. Cet exercice m'a permis de (re)découvrir une fonction que je n'utilise que trop rarement au quotidien, le foldLeft (et son corollaire, le foldRight). 
+Ces méthodes, qui permettent de parcourir des listes effectuant des opération entre la valeur courante et la valeur précédente tout en permettant de retourner un type complètement différent, méritent qu'on s'y attarde un peu.
+Intéressons-nous tout d'abord à la signature de cette méthode
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:10]:
```suggestion
Intéressons-nous tout d'abord à la signature de cette méthode :
```

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+[source, scala]
+----
+def foldLeft[B](z: B)(op: (B, A) => B): B
+----
+Nous voyons ici plusieurs choses intéressantes :
+1. La fonction est typée B, qui est le type de sortie
+2. Elle prend en entrée un premier paramètre z (zéro) de ce fameux type B. Ce paramètre est important car c'est le point de départ de la fonction, il sert à l'initialiser et à collecter les résultats. Ce sera notre accumulateur
+3. Et en second paramètre "op" (opération) une fonction qui prend deux paramètres (B, le type de sortie et A , le type des éléments de la liste) et qui permet d'obtenir ce fameux type de sortie B.
+En gros, cette fonction va parcourir tous les éléments de la liste [A] et va collecter le résultat dans l'accumulateur [B]. A la fin, on se retrouve avec [B].
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:19]:
```suggestion
Pour résumer, cette fonction va parcourir tous les éléments de la liste [A] et va collecter le résultat dans l'accumulateur [B]. A la fin, on se retrouve avec [B].
```

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+Commençons par un exemple simple. Nous allons additionner chaque chiffre d'une liste avec son prédécesseur pour obtenir le résultat final.
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:20]:
```suggestion
Commençons par un exemple simple. Nous allons additionner chaque chiffre d'une liste avec son prédécesseur pour obtenir la somme de cette liste.
```

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+val numbers = List(1, 2, 3, 4)
+numbers.reduce(_ + _)
+// Ce qui equivaut à
+numbers.sum
+// Result => Out: Int = 10
+----
+Rapprochons-nous petit à petit de cas plus concrets.
+Dans l'exemple suivant, je voudrais, non pas faire la somme des chiffres de la liste, mais obtenir une nouvelle liste dont chaque chiffre serait l'addition du chiffre de la liste "numbers" avec son prédécesseur.
+Nous allons parcourir la liste "List(1, 2, 3, 4)" et à chaque itération, faire cette opération :
+1. accumulateur vide et chiffre courant = 1 : résultat ==> accumulateur = List(1)
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:44]: Peut être ajouté du formatting sur ces points ?

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+----
+acc.andThen { players =>
+            current.map(_ +: players)
+          }
+----
+Là encore, il s'agit de la version Cats de left.map(...)
+Enfin, nous enveloppons notre bloc de `EitherT.fromEither[Future] { { ... }.toEither }` pour passer du type `Validated` au type `EitherT`
+
+== Conclusion
+Comme nous l'avons vu, foldLeft et foldRight sont un peu les boîtes à tout faire lorsqu'on veut parcourir une liste dans un ordre donné et comparer les entrées entre elles tout en retournant un type différent au besoin. J'espère que cet article vous aura éclairé sur la manière de les utiliser simplement et vous aura convaincu d'en user et même d'en abuser !
```
Comment (VincentBrule) [inline posts/2022-06-27-foldleft-introduction.adoc:277]:
```suggestion
Comme nous l'avons vu, *foldLeft* et *foldRight* sont un peu les boîtes à tout faire lorsqu'on veut parcourir une liste dans un ordre donné et comparer les entrées entre elles tout en retournant un type différent au besoin. J'espère que cet article vous aura éclairé sur la manière de les utiliser simplement et vous aura convaincu d'en user et même d'en abuser !
```

File: posts/2022-06-27-foldleft-introduction.adoc (hunk 1 above)
Comment (ggarvanese) [reply posts/2022-06-27-foldleft-introduction.adoc:1]: Ah ! je n'avais pas remarqué qu'on pouvait afficher le résultat final de cette manière. Merci !

File: posts/2022-06-27-foldleft-introduction.adoc [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+4. accumulateur (1, 3, 6) et chiffre courant = 4 : résultat ==> accumulateur = List(1, 3, 6, 4+6)
+Résultat final : `newList = List(1, 3, 6, 10)``
+Une écriture possible peut s'appuyer sur un pattern matching avec une condition dédiée à la condition initiale, c'est-à-dire lorsque nous démarrons avec une liste vide car nous aurions une erreur à l'exécution si nous appelions `accumulator.last` sur une liste vide. Comme nous n'avons rien à additionner au début, nous remplissons simplement la liste avec le premier chiffre.
+[source, scala]
+----
+val numbers = List(1, 2, 3, 4)
+val newList: Seq[Int] = numbers.foldLeft[Seq[Int]](Seq.empty) {
+    case (accumulator, number) if accumulator.isEmpty => accumulator :+ number
+    case (accumulator, number) => accumulator :+ (accumulator.last + number)
+}
```
Comment (eloots) [inline posts/2022-06-27-foldleft-introduction.adoc:56]: Guillaume, I would add a remark here about performance/inefficiency of the shown implementation.

`foldLeft[Seq[Int]](Seq.empty) ...` is generic in the sense that it takes a `Seq[Int]`. As you append items to the accumulator (initialised as `Seq.empty[Int]`), this is very inefficient as the compiler will pick `List` as the collection type for the accumulator. The inefficiency presents itself twice in each step of processing the elements in the input sequence: first in getting the last element in the sequence (`accumulator.last`) and secondly when appending the new element at the end of the sequence (`accumulator :+ ...`).
I think you should point this out in the post and you can refer to the [Scala Collection Performance Characteristics](https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html).
I think doing so gives you the possibility to give an alternative solution that, while staying generic, avoids taking the performance hit. It would look like this:
```
List(1,2,3,4).foldLeft(Seq.empty[Int]) {
  case (Nil, el) => Seq(el)
  case (accumulator, el) => (accumulator.head + el) +: accumulator
}
```

In this case, getting the `head` element and prepending a new element to the accumulator are very fast (constant time) operations. Of course, this implementation doesn't produce the same result as the original implementation as the element order is reversed. This can be solved either by using a `foldRight` version of the implementation or calling `reverse` on the `foldLeft` based implementation.

It is interesting to point out that `foldRight` in the Scala Collection library is `foldLeft` in disguise:

From [scala/collection/IterableOnce.scala](https://github.com/scala/scala/blob/v2.13.8/src/library/scala/collection/IterableOnce.scala#L309):

```
def foldRight[B](z: B)(op: (A, B) => B): B = reversed.foldLeft(z)((b, a) => op(a, b))
```

Comment (ggarvanese) [reply posts/2022-06-27-foldleft-introduction.adoc:56]: Thank you Eric for this very interesting point 👍  
As I'm not very comfortable with this topic (especially to popularise it), would you mind if I paraphrased your comment in a section crediting you? I would also add your link as a source

File: posts/2022-06-27-foldleft-introduction.adoc (hunk 7 above)
Comment (ggarvanese) [reply posts/2022-06-27-foldleft-introduction.adoc:44]: Je ne suis pas sûr de ce que tu as en tête. J'ai revu la ponctuation pour apporter un peu de clarté

File: posts/2022-06-27-foldleft-introduction.adoc (hunk 9 above)
Comment (eloots) [reply posts/2022-06-27-foldleft-introduction.adoc:77]: Guillaume, yes, feel free to add it to the blog article (and no need to mention me btw) !

**Review by eloots — APPROVED** [review]
👍🏼

File: posts/2022-06-27-foldleft-introduction-en.adoc [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+        .toEither
+    }
+  } yield updatedPlayers
+Await.result(resultEitherT.value, 1.second)
+/* Result =>
+Left(java.lang.IllegalArgumentException: Unable to delete the task due to status is Enrolled)
+*/
+----
+
+Précisions :
```
Comment (eloots) [inline posts/2022-06-27-foldleft-introduction-en.adoc:317]: (still in French) maybe change it to:

Attentive users may have noticed this piece of code:

File: posts/2022-06-27-foldleft-introduction-en.adoc [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+The flatMap allows you to access the Player Sequence located in the Right of the accumulator's Either and return an `Either[Exception, Seq[Player]]` instead of an `Either[Exception, Either[Exception, Seq[Player]]`.
+
+[source, scala]
+----
+.left
+.map(error => ... 
+----
+If there is no Right, then Left is considered a return type. As there is only one possible Left in our return type Either[Exception, Seq[Player]], then processing is stopped as soon as it is filled in.
+
+With the Cats library, the result would be as follows :
```
Comment (eloots) [inline posts/2022-06-27-foldleft-introduction-en.adoc:265]: With the Cats library, we can code this up as follows:

File: posts/2022-06-27-foldleft-introduction-en.adoc [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+
+[source, scala]
+----
+acc.andThen { players =>
+            current.map(_ +: players)
+          }
+----
+
+Again, this is the Cats version of left.map(...)
+Finally, we wrap our block with `EitherT.fromEither[Future] { { ... }.toEither }` to change the type from `Validated` to `EitherT`.
```
Comment (eloots) [inline posts/2022-06-27-foldleft-introduction-en.adoc:327]: This is the Cats version of left.map(...)

Finally, we wrap our block with `EitherT.fromEither[Future] { { ... }.toEither }` to change the type from `Validated` to `EitherT`.

Note that we have used `EitherT` and `Validated`, two Cats specific types.

File: posts/2022-06-27-foldleft-introduction-en.adoc [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+}.reverse
+----
+
+In this way, accessing or adding an element to the accumulator will be a constant time (fast) operation. But as this implementation produces a reverse result, it is sufficient to add a `.reverse` at the end, or to use a foldRight, which you will notice is just a reverse foldLeft.
+
+`def foldRight[B](z: B)(op: (A, B) => B): B = reversed.foldLeft(z)((b, a) => op(a, b))`
+
+== Conclusion
+
+As we have seen, *foldLeft* and *foldRight* are kind of all-purpose boxes when you want to go through a list in a given order and compare the entries with each other while returning a different type if needed.
```
Comment (eloots) [inline posts/2022-06-27-foldleft-introduction-en.adoc:370]: Maybe you can rephrase this as:

As demonstrated in this article, *foldLeft* and *foldRight* are very powerful methods that can be considered the equivalent of the Swiss Army knife in the Scala collection library: it operates on a collection of elements of some type A and can generate a value that is the same type A or a completely different type B.
